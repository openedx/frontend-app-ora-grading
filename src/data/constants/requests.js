import { StrictDict } from '@src/utils';

export const RequestStates = StrictDict({
  inactive: 'inactive',
  pending: 'pending',
  completed: 'completed',
  failed: 'failed',
});

export const RequestKeys = StrictDict({
  batchUnlock: 'batchUnlock',
  downloadFiles: 'downloadFiles',
  fetchSubmission: 'fetchSubmission',
  fetchSubmissionStatus: 'fetchSubmissionStatus',
  initialize: 'initialize',
  prefetchNext: 'prefetchNext',
  prefetchPrev: 'prefetchPrev',
  setLock: 'setLock',
  submitGrade: 'submitGrade',
});

export const ErrorCodes = StrictDict({
  missingParam: 'ERR_MISSING_PARAM',
});

export const ErrorStatuses = StrictDict({
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  serverError: 500,
});
