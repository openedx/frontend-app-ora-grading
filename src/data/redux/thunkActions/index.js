import { StrictDict } from '@src/utils';
// eslint-disable-next-line import/no-cycle
import app from './app';
import download from './download';
import grading from './grading';

export default StrictDict({
  app,
  download,
  grading,
});
