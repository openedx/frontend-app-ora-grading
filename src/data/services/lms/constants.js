import { StrictDict } from '@src/utils';

export const lockStatuses = StrictDict({
  unlocked: 'unlocked',
  locked: 'locked',
  inProgress: 'in-progress',
});

export const gradeStatuses = StrictDict({
  ungraded: 'ungraded',
  graded: 'graded',
});

export const gradingStatuses = StrictDict({
  ungraded: gradeStatuses.ungraded,
  graded: gradeStatuses.graded,
  locked: lockStatuses.locked,
  inProgress: lockStatuses.inProgress,
});

export const feedbackRequirement = StrictDict({
  disabled: 'disabled',
  required: 'required',
  optional: 'optional',
});

export const fileUploadResponseOptions = StrictDict({
  required: 'required',
  optional: 'optional',
  none: 'none',
});

export const paramKeys = StrictDict({
  oraLocation: 'oraLocation',
  submissionUUID: 'submissionUUID',
});

export const oraTypes = StrictDict({
  team: 'team',
  individual: 'individual',
});

export const submissionFields = StrictDict({
  dateSubmitted: 'dateSubmitted',
  gradingStatus: 'gradingStatus',
  score: 'score',
  teamName: 'teamName',
  username: 'username',
});
