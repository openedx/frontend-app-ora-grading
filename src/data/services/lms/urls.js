import { StrictDict } from '@src/utils';
import { getConfig } from '@edx/frontend-platform';

const baseUrl = () => getConfig().LMS_BASE_URL;

const api = () => `${baseUrl()}/api/`;
const baseEsgUrl = () => `${api()}ora_staff_grader/`;

const oraInitializeUrl = () => `${baseEsgUrl()}initialize`;
const fetchSubmissionUrl = () => `${baseEsgUrl()}submission`;
const fetchSubmissionFilesUrl = () => `${baseEsgUrl()}submission/files`;
const fetchSubmissionStatusUrl = () => `${baseEsgUrl()}submission/status`;
const fetchSubmissionLockUrl = () => `${baseEsgUrl()}submission/lock`;
const batchUnlockSubmissionsUrl = () => `${baseEsgUrl()}submission/batch/unlock`;
const updateSubmissionGradeUrl = () => `${baseEsgUrl()}submission/grade`;

const course = (courseId) => `${baseUrl()}/courses/${courseId}`;

const openResponse = (courseId) => (
  `${course(courseId)}/instructor#view-open_response_assessment`
);
const ora = (courseId, locationId) => `${course(courseId)}/jump_to/${locationId}`;

export default StrictDict({
  api,
  oraInitializeUrl,
  fetchSubmissionUrl,
  fetchSubmissionFilesUrl,
  fetchSubmissionStatusUrl,
  fetchSubmissionLockUrl,
  batchUnlockSubmissionsUrl,
  updateSubmissionGradeUrl,
  baseUrl,
  course,
  openResponse,
  ora,
});
