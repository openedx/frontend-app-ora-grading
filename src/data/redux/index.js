import { combineReducers } from 'redux';

import { StrictDict } from '@src/utils';

import * as app from './app';
import * as grading from './grading';
import * as requests from './requests';
import * as submissions from './submissions';

export { default as thunkActions } from './thunkActions';

const modules = {
  app,
  grading,
  requests,
  submissions,
};

const moduleProps = (propName) => Object.keys(modules).reduce(
  (obj, moduleKey) => ({ ...obj, [moduleKey]: modules[moduleKey][propName] }),
  {},
);

const rootReducer = combineReducers(moduleProps('reducer'));

const actions = StrictDict(moduleProps('actions'));

const selectors = StrictDict(moduleProps('selectors'));

export { actions, selectors };

export default rootReducer;
