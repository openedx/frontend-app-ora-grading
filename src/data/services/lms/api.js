import { StrictDict } from 'utils';
import { locationId } from 'data/constants/app';
import { paramKeys } from './constants';
import urls from './urls';
import {
  client,
  get,
  post,
  stringifyUrl,
} from './utils';

/*********************************************************************************
 * GET Actions
 *********************************************************************************/

/**
 * get('/api/initialize', { oraLocation })
 * @return {
 *   oraMetadata: { name, prompt, type ('individual' vs 'team'), rubricConfig, fileUploadResponseConfig },
 *   courseMetadata: { courseOrg, courseName, courseNumber, courseId },
 *   submissions: {
 *     [submissionUUID]: {
 *       id: <submissionUUID>, (not currently used)
 *       username
 *       submissionUUID
 *       dateSubmitted (timestamp)
 *       gradeStatus (['ungraded', 'graded', 'locked', 'locked_by_you'?])
 *       grade: { pointsEarned, pointsPossible }
 *     },
 *     ...
 *   },
 * }
 */
const initializeApp = () => get(
  stringifyUrl(urls.oraInitializeUrl(), {
    [paramKeys.oraLocation]: locationId(),
  }),
).then(response => response.data);

/**
 * get('/api/submission', { oraLocation, submissionUUID })
 * @return {
 *   submission: {
 *     gradeData,
 *     gradeStatus,
 *     response: { files: [{}], text: <html> },
 *   },
 * }
 */
const fetchSubmission = (submissionUUID) => get(
  stringifyUrl(urls.fetchSubmissionUrl(), {
    [paramKeys.oraLocation]: locationId(),
    [paramKeys.submissionUUID]: submissionUUID,
  }),
).then(response => ({
  ...response.data,
  response: {
    ...response.data.response,
    // The API includes metadata for deleted files (without a download url);
    // we don't want to deal with deleted files here, so filter them out like the student view does.
    files: response.data.response.files.filter((file) => file.downloadUrl),
  },
}));

/**
 * get('/api/submission/files', { oraLocation, submissionUUID })
 * @return {
 *     response: { files: [{}], text: <html> },
 * }
 */
const fetchSubmissionFiles = (submissionUUID) => get(
  stringifyUrl(urls.fetchSubmissionFilesUrl(), {
    [paramKeys.oraLocation]: locationId(),
    [paramKeys.submissionUUID]: submissionUUID,
  }),
).then(response => ({
  ...response.data,
  // The API includes metadata for deleted files (without a download url);
  // we don't want to deal with deleted files here, so filter them out like the student view does.
  files: response.data.files.filter((file) => file.downloadUrl),
}));

/**
 * fetches the current grade, gradeStatus, and rubricResponse data for the given submission
 * get('/api/submissionStatus', { oraLocation, submissionUUID })
 *   @return {obj} submissionStatus object
 *   {
 *     gradeData,
 *     gradeStatus,
 *     lockStatus,
 *   }
 */
const fetchSubmissionStatus = (submissionUUID) => get(
  stringifyUrl(urls.fetchSubmissionStatusUrl(), {
    [paramKeys.oraLocation]: locationId(),
    [paramKeys.submissionUUID]: submissionUUID,
  }),
).then(response => response.data);

/**
 * post('api/lock', { oraLocation, submissionUUID });
 * @param {string} submissionUUID
 */
const lockSubmission = (submissionUUID) => post(
  stringifyUrl(urls.fetchSubmissionLockUrl(), {
    [paramKeys.oraLocation]: locationId(),
    [paramKeys.submissionUUID]: submissionUUID,
  }),
).then(response => response.data);

/**
 * unlockSubmission(submissionUUID)
 * @param {string} submissionUUID
 */
const unlockSubmission = (submissionUUID) => client().delete(
  stringifyUrl(urls.fetchSubmissionLockUrl(), {
    [paramKeys.oraLocation]: locationId(),
    [paramKeys.submissionUUID]: submissionUUID,
  }),
).then(response => response.data);

/**
 * batchUnlockSubmissions(submissionUUIDs)
 * @param {string[]} submissionUUIDs - list of submission uuids
 */
const batchUnlockSubmissions = (submissionUUIDs) => post(
  stringifyUrl(
    urls.batchUnlockSubmissionsUrl(),
    { [paramKeys.oraLocation]: locationId() },
  ),
  { submissionUUIDs },
).then(response => response.data);

/*
 * post('api/updateGrade', { submissionUUID, gradeData })
 * @param {object} gradeData - full grading submission data
 */
const updateGrade = (submissionUUID, gradeData) => post(
  stringifyUrl(urls.updateSubmissionGradeUrl(), {
    [paramKeys.oraLocation]: locationId(),
    [paramKeys.submissionUUID]: submissionUUID,
  }),
  gradeData,
).then(response => response.data);

export default StrictDict({
  initializeApp,
  fetchSubmission,
  fetchSubmissionFiles,
  fetchSubmissionStatus,
  lockSubmission,
  updateGrade,
  unlockSubmission,
  batchUnlockSubmissions,
});
