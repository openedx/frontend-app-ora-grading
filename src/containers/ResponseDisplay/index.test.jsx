import { MathJaxContext } from 'better-react-mathjax';

import { render, screen } from '@testing-library/react';
import { fileUploadResponseOptions } from 'data/services/lms/constants';
import { selectors } from 'data/redux';
import { ResponseDisplay, mapStateToProps } from '.';

jest.mock('data/redux', () => ({
  selectors: {
    grading: {
      selected: {
        response: jest.fn((state) => state.response || { text: [], files: [] }),
      },
    },
    app: {
      ora: {
        fileUploadResponseConfig: jest.fn((state) => state.fileUploadResponseConfig || 'optional'),
      },
    },
  },
}));

jest.mock('./SubmissionFiles', () => jest.fn(({ files }) => (
  <div data-testid="submission-files">Files: {files.length}</div>
)));

jest.mock('./PreviewDisplay', () => jest.fn(({ files }) => (
  <div data-testid="preview-display">Preview: {files.length}</div>
)));

jest.mock('dompurify', () => () => ({
  sanitize: (text) => text,
}));

jest.mock('html-react-parser', () => (text) => text);

describe('ResponseDisplay', () => {
  const defaultProps = {
    response: {
      text: ['some text response here', 'another text response'],
      files: [
        {
          name: 'some file name.jpg',
          description: 'description for the file',
          downloadURL: '/valid-url-wink-wink',
        },
        {
          name: 'file number 2.jpg',
          description: 'description for this file',
          downloadURL: '/url-2',
        },
      ],
    },
    fileUploadResponseConfig: 'optional',
  };

  beforeAll(() => {
    global.window = {};
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('behavior', () => {
    it('renders response display container', () => {
      const { container } = render(
        <MathJaxContext>
          <ResponseDisplay {...defaultProps} />
        </MathJaxContext>,
      );
      const responseDisplay = container.querySelector('.response-display');
      expect(responseDisplay).toBeInTheDocument();
    });

    it('displays text content in cards', () => {
      const { container } = render(
        <MathJaxContext>
          <ResponseDisplay {...defaultProps} />
        </MathJaxContext>,
      );
      const textContents = container.querySelectorAll('.response-display-text-content');
      expect(textContents).toHaveLength(defaultProps.response.text.length);
      expect(textContents[0]).toHaveTextContent('some text response here');
      expect(textContents[1]).toHaveTextContent('another text response');
    });

    it('displays submission files when file upload is allowed', () => {
      render(
        <MathJaxContext>
          <ResponseDisplay {...defaultProps} />
        </MathJaxContext>,
      );
      const submissionFiles = screen.getByTestId('submission-files');
      expect(submissionFiles).toBeInTheDocument();
      expect(submissionFiles).toHaveTextContent('Files: 2');
    });

    it('displays preview display when file upload is allowed', () => {
      render(
        <MathJaxContext>
          <ResponseDisplay {...defaultProps} />
        </MathJaxContext>,
      );
      const previewDisplay = screen.getByTestId('preview-display');
      expect(previewDisplay).toBeInTheDocument();
      expect(previewDisplay).toHaveTextContent('Preview: 2');
    });

    it('does not display file components when file upload is disabled', () => {
      render(
        <MathJaxContext>
          <ResponseDisplay {...defaultProps} fileUploadResponseConfig={fileUploadResponseOptions.none} />
        </MathJaxContext>,
      );
      expect(screen.queryByTestId('submission-files')).not.toBeInTheDocument();
      expect(screen.queryByTestId('preview-display')).not.toBeInTheDocument();
    });

    it('renders empty content when no text response provided', () => {
      const { container } = render(
        <MathJaxContext>
          <ResponseDisplay {...defaultProps} response={{ text: [], files: [] }} />
        </MathJaxContext>,
      );
      const textContents = container.querySelectorAll('.response-display-text-content');
      expect(textContents).toHaveLength(0);
    });
  });

  describe('mapStateToProps', () => {
    const testState = {
      response: {
        text: ['test text'],
        files: ['file1', 'file2'],
      },
      fileUploadResponseConfig: 'required',
    };

    it('maps response from grading.selected.response selector', () => {
      const mapped = mapStateToProps(testState);
      expect(mapped.response).toEqual(selectors.grading.selected.response(testState));
    });

    it('maps fileUploadResponseConfig from app.ora.fileUploadResponseConfig selector', () => {
      const mapped = mapStateToProps(testState);
      expect(mapped.fileUploadResponseConfig).toEqual(selectors.app.ora.fileUploadResponseConfig(testState));
    });
  });
});
