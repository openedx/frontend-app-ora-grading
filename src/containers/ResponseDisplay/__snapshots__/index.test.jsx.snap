// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponseDisplay component snapshot file upload disable with valid response 1`] = `
<div
  className="response-display"
>
  <MathJax>
    <Card
      key="0"
    >
      <Card.Section
        className="response-display-text-content"
        data-testid="response-display-text-content"
      >
        parsed html (sanitized (some text response here))
      </Card.Section>
    </Card>
  </MathJax>
</div>
`;

exports[`ResponseDisplay component snapshot file upload disabled without response 1`] = `
<div
  className="response-display"
>
  <SubmissionFiles
    data-testid="submission-files"
    files={[]}
  />
  <PreviewDisplay
    data-testid="allow-file-upload"
    files={[]}
  />
</div>
`;

exports[`ResponseDisplay component snapshot file upload enable with valid response 1`] = `
<div
  className="response-display"
>
  <SubmissionFiles
    data-testid="submission-files"
    files={
      [
        {
          "description": "description for the file",
          "downloadURL": "/valid-url-wink-wink",
          "name": "some file name.jpg",
        },
        {
          "description": "description for this file",
          "downloadURL": "/url-2",
          "name": "file number 2.jpg",
        },
      ]
    }
  />
  <PreviewDisplay
    data-testid="allow-file-upload"
    files={
      [
        {
          "description": "description for the file",
          "downloadURL": "/valid-url-wink-wink",
          "name": "some file name.jpg",
        },
        {
          "description": "description for this file",
          "downloadURL": "/url-2",
          "name": "file number 2.jpg",
        },
      ]
    }
  />
  <MathJax>
    <Card
      key="0"
    >
      <Card.Section
        className="response-display-text-content"
        data-testid="response-display-text-content"
      >
        parsed html (sanitized (some text response here))
      </Card.Section>
    </Card>
  </MathJax>
</div>
`;

exports[`ResponseDisplay component snapshot file upload enable without response 1`] = `
<div
  className="response-display"
>
  <SubmissionFiles
    data-testid="submission-files"
    files={[]}
  />
  <PreviewDisplay
    data-testid="allow-file-upload"
    files={[]}
  />
</div>
`;
